pkg/customresource/registry_factory.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2021 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package customresource
+
+import (
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+
+	generator "k8s.io/kube-state-metrics/v2/pkg/metric_generator"
+)
+
+// RegistryFactory is a registry interface for a CustomResourceStore.
+// Users who want to extend the kube-state-metrics to support Custom Resource metrics should
+// implement this interface.
+type RegistryFactory interface {
+	// Name returns the name of custom resource.
+	//
+	// Example:
+	//
+	// func (f *FooFactory) Name() string {
+	//	return "foos"
+	// }
+	Name() string
+
+	// CreateClient creates a new custom resource client for the given config.
+	//
+	// Example:
+	//
+	// func (f *FooFactory) CreateClient(cfg *rest.Config) (interface{}, error) {
+	// 	return clientset.NewForConfig(cfg)
+	// }
+	CreateClient(cfg *rest.Config) (interface{}, error)
+
+	// MetricFamilyGenerators returns the metric family generators to generate metric families with a
+	// Kubernetes custom resource object.
+	//
+	// Example:
+	//
+	// func (f *FooFactory) MetricFamilyGenerators(allowAnnotationsList, allowLabelsList []string) []generator.FamilyGenerator {
+	//	return []generator.FamilyGenerator{
+	//		*generator.NewFamilyGenerator(
+	//			"kube_foo_spec_replicas",
+	//			"Number of desired replicas for a foo.",
+	//			metric.Gauge,
+	//			"",
+	//			wrapFooFunc(func(f *samplev1alpha1.Foo) *metric.Family {
+	//				return &metric.Family{
+	//					Metrics: []*metric.Metric{
+	//						{
+	//							Value: float64(*f.Spec.Replicas),
+	//						},
+	//					},
+	//				}
+	//			}),
+	//		),
+	//		*generator.NewFamilyGenerator(
+	//			"kube_foo_status_replicas_available",
+	//			"The number of available replicas per foo.",
+	//			metric.Gauge,
+	//			"",
+	//			wrapFooFunc(func(f *samplev1alpha1.Foo) *metric.Family {
+	//				return &metric.Family{
+	//					Metrics: []*metric.Metric{
+	//						{
+	//							Value: float64(f.Status.AvailableReplicas),
+	//						},
+	//					},
+	//				}
+	//			}),
+	//		),
+	//	}
+	// }
+	MetricFamilyGenerators(allowAnnotationsList, allowLabelsList []string) []generator.FamilyGenerator
```

Comment:
Should the implementation of `(customresource.RegistryFactory).MetricFamilyGenerators` be responsible for filtering labels and annotations based on which are allowed? What are the risks of doing this?

- What does this do to the expectation that `--metric-labels-allowlist` and `--metric-annotations-allowlist` work the same way for all metrics?

- Does this affect the flexibility in terms of adding onto these labels- and annotation allow lists with other types of lists in the future? With backward compatibility for the implementation of `customresource.RegistryFactory` in mind.

Reply:
> What does this do to the expectation that --metric-labels-allowlist and --metric-annotations-allowlist work the same way for all metrics?
I don't figure out any good ways to keep the custom resource registry working the same way. Some metrics do not have these labels and some do focusing. It is really hard so I just let the interface implementers do this. Do you have any ideas?

> Does this affect the flexibility in terms of adding onto these labels- and annotation allow lists with other types of lists in the future? With backward compatibility for the implementation of customresource.RegistryFactory in mind.
How about using a `...[]string`? Like `MetricFamilyGenerators(allowanceLists ...[]string) []generator.FamilyGenerator`. It's a lit bit weird😂.

Reply:
I personally think that to do this properly and prevent future headaches that a refactor of the way labels and annotations are filtered is appropriate here. If it is possible to refactor these filters outside of the `(generator.FamilyGenerator).GenerateFunc` that would make the generate function by definition less complex and (in the most optimal case) there is only one place where labels and annotations are filtered.

Initially I went with the idea of doing this **after** the metrics were generated but then I realized that `--metric-labels-allowlist` and `--metric-annotations-allowlist` are depended on whether the given label on the metric **was** either a label or annotation on the original resource object.

I then went looking and found the function below which eventually calls the `(metricsstore.MetricsStore).generateMetricsFuncs` with the original resource object.

https://github.com/kubernetes/kube-state-metrics/blob/840d30ce57f75d0ec466a2d612879602c433d5e6/pkg/metrics_store/metrics_store.go#L63-L82

I might be completely insane obviously so I'd like to know what the others think about this.

**^ all of the above is said with the intent of making this refactor a separate PR by the way, who might be you, someone else or myself** 😉

Reply:
Wow, I think it's a pretty good idea! Users don't need to handle the associated label and annotation filter in their own codes. So how about fuction like `MetricFamilyGenerators() []generator.FamilyGenerator`?

Reply:
I'm sorry, I think I've completely mistaken what `--metric-labels-allowlist` and `--metric-annotations-allowlist` are for and thus "recommended" something that is completely unnecessary.

`--metric-labels-allowlist` and `--metric-annotations-allowlist` are for `_annotations` and `_labels` metrics (e.g. `kubernetes_node_labels` and `kubernetes_node_annotations`).

There is no point in refactoring this whatsoever and the implementation is indeed responsible for using the label and annotation filters appropriately. If a CR metric is added and a similar `_labels` or `_annotations` metric for this CR is added the implementation is obviously responsible for taking the label and annotation allowlist into account.

Please ignore my clumsiness 😆

Reply:
Ops, that's OK. I didn't realize too 😂